### PR TITLE
Add explicit errors when google returns non-OK status

### PR DESCRIPTION
**Pull request by kiote**
before this we were returning "exceeded query limit" error on any non-OK status from Google, which looks confusing when we are trying to deal with this errors in our app.

  This change add separate error classes which simply returns response from Google if we had non-OK status.

Some tests failed on travis, but it's Rubocop only and non-related to new code (except class length).

### CODE REVIEW

File: lib/google_maps_geocoder/google_maps_geocoder.rb [hunk 1]
```diff
@@ -3,6 +3,8 @@
 require 'net/http'
 require 'rack'
 
+require 'google_maps_geocoder/google_maps_geocoder_error_handler'
```
Comment (kiote) [inline lib/google_maps_geocoder/google_maps_geocoder.rb]: not sure what do you think about adding new classes here, since before it was just one class, but the length of this class has started to be a problem for rubocop

File: lib/google_maps_geocoder/google_maps_geocoder.rb [hunk 2]
```diff
@@ -96,11 +97,28 @@ def http(uri)
...
     response = http(uri).request(Net::HTTP::Get.new(uri.request_uri))
     ActiveSupport::JSON.decode response.body
   end
 
+  def handle_exception(google_response_json)
+    status = google_response_json['status']
+    # no exception
+    return if status == 'OK'
+
+    message = GoogleMapsGeocoderErrorHandler::GeneralGoecodingError\
```
Comment (kiote) [inline lib/google_maps_geocoder/google_maps_geocoder.rb]: this module name seems to be too long as well, but having just a `GoogleMapsGeocoder::GeneralGoecodingError` involves some fundamental changes here, so I prefer to wait for your response first.

**Comment by ivanoblomov** [comment]
This is a good idea. Also, don't worry too much about the Coveralls warnings. I was always skeptical of their three decimals of precision.

File: lib/google_maps_geocoder/google_maps_geocoder.rb (hunk 1 above)
Comment (ivanoblomov) [inline lib/google_maps_geocoder/google_maps_geocoder.rb]: Let's rename `google_maps_geocoder_error_handler` to `geocoding_exception`. Shorter and sweeter.

File: lib/google_maps_geocoder/google_maps_geocoder.rb (hunk 2 above)
Comment (ivanoblomov) [inline lib/google_maps_geocoder/google_maps_geocoder.rb]: Similarly, rename this to `GoogleMapsGeocoder::GoecodingException`. I like the distinction Java made between error (unrecoverable) and exception (recoverable, meant to be caught).

File: lib/google_maps_geocoder/google_maps_geocoder.rb (hunk 1 above)
Comment (ivanoblomov) [inline lib/google_maps_geocoder/google_maps_geocoder.rb]: And go ahead and keep it all in one class file. You can disable the Rubocop warning by adding this before the class declaration:

```ruby
# rubocop:disable Metrics/ClassLength
class GoogleMapsGeocoder
```

File: lib/google_maps_geocoder/google_maps_geocoder.rb (hunk 2 above)
Comment (ivanoblomov) [inline lib/google_maps_geocoder/google_maps_geocoder.rb]: Actually, let's stick to Ruby's conventions according to http://blog.honeybadger.io/ruby-exception-vs-standarderror-whats-the-difference/.

TL;DR: Call it `GoogleMapsGeocoder::GoecodingError`, have it extend `StandardError`, and, if necessary, rescue either class—*not* `Exception`.

File: lib/google_maps_geocoder/google_maps_geocoder.rb (hunk 1 above)
Comment (ivanoblomov) [inline lib/google_maps_geocoder/google_maps_geocoder.rb]: Note also that you can automatically fix most Rubocop errors with:

```bash
bundle exec rubocop -a
```

File: lib/google_maps_geocoder/google_maps_geocoder.rb [hunk 3]
```diff
@@ -52,8 +54,7 @@ class GoogleMapsGeocoder
   #   chez_barack = GoogleMapsGeocoder.new '1600 Pennsylvania Ave'
   def initialize(data)
     @json = data.is_a?(String) ? json_from_url(data) : data
-    fail "Geocoding \"#{data}\" exceeded query limit! Google returned...\n"\
-         "#{@json.inspect}" if @json.blank? || @json['status'] != 'OK'
+    handle_exception(@json)
```
Comment (ivanoblomov) [inline lib/google_maps_geocoder/google_maps_geocoder.rb]: Please rename the method and only call it if you need to:

```ruby
handle_error if @json.blank? || @json['status'] != 'OK'
```

 Note you *don't* have to pass in the JSON as a method argument because `@json` is an instance variable.

File: lib/google_maps_geocoder/google_maps_geocoder.rb [hunk 4]
```diff
@@ -96,11 +97,28 @@ def http(uri)
...
+    # no exception
+    return if status == 'OK'
+
+    message = GoogleMapsGeocoderErrorHandler::GeneralGoecodingError\
+              .new(google_response_json).message
+
+    # for status codes see https://developers.google.com/maps/documentation/geocoding/intro#StatusCodes
+    raise GoogleMapsGeocoderErrorHandler::ZeroResultsError,\
+          message if status == 'ZERO_RESULTS'
+    raise GoogleMapsGeocoderErrorHandler::QueryLimitError,\
```
Comment (ivanoblomov) [inline lib/google_maps_geocoder/google_maps_geocoder.rb]: I like your error-class names here, but do we really need the parent class `GoogleMapsGeocoderErrorHandler`? I'd rather extend `StandardError` directly as mentioned earlier.

File: lib/google_maps_geocoder/google_maps_geocoder_error_handler.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  class GeneralGoecodingError < StandardError
+    def initialize(response_json = '')
+      @json = response_json
+      super
+    end
+
+    def message
+      "Google returned:\n#{@json.inspect}"
+    end
+  end
```
Comment (ivanoblomov) [inline lib/google_maps_geocoder/google_maps_geocoder_error_handler.rb]: Ah, now I see why you wanted the superclass. I'm good with this, but please shorten `GeneralGoecodingError` to `GeocodingError`.

Also, is the module necessary? Can't we have the error classes in the same level and file as `GoogleMapsGeocoder`?

File: lib/google_maps_geocoder/google_maps_geocoder.rb [hunk 6]
```diff
@@ -1,18 +1,43 @@
 require 'active_support'
+require 'active_support/core_ext/string/inflections'
```
Comment (kiote) [inline lib/google_maps_geocoder/google_maps_geocoder.rb]: I'll need this later to have less repetitive code (with help of constantize)

File: lib/google_maps_geocoder/google_maps_geocoder.rb [hunk 7]
```diff
@@ -1,18 +1,43 @@
...
+    def message
+      "Google returned:\n#{@json.inspect}"
+    end
+  end
+
+  class ZeroResultsError < GeocodingError; end
+  class QueryLimitError < GeocodingError; end
+  class ReqestDeniedError < GeocodingError; end
+  class InvalidRequestError < GeocodingError; end
+  class UnknownError < GeocodingError; end
```
Comment (kiote) [inline lib/google_maps_geocoder/google_maps_geocoder.rb]: with this set of classes we cover all situations google provided as "not okay"

File: lib/google_maps_geocoder/google_maps_geocoder.rb [hunk 8]
```diff
@@ -54,7 +79,7 @@ class GoogleMapsGeocoder
   #   chez_barack = GoogleMapsGeocoder.new '1600 Pennsylvania Ave'
   def initialize(data)
     @json = data.is_a?(String) ? json_from_url(data) : data
-    handle_exception(@json)
+    handle_error if @json.blank? || @json['status'] != 'OK'
```
Comment (kiote) [inline lib/google_maps_geocoder/google_maps_geocoder.rb]: renamed and added a condition as you suggested to

File: lib/google_maps_geocoder/google_maps_geocoder.rb [hunk 9]
```diff
@@ -104,19 +129,16 @@ def json_from_url(url)
...
-    return if status == 'OK'
-
-    message = GoogleMapsGeocoderErrorHandler::GeneralGoecodingError\
-              .new(google_response_json).message
+  def handle_error
+    status = @json['status']
+    message = GeocodingError.new(@json).message
 
     # for status codes see https://developers.google.com/maps/documentation/geocoding/intro#StatusCodes
-    raise GoogleMapsGeocoderErrorHandler::ZeroResultsError,\
```
Comment (kiote) [inline lib/google_maps_geocoder/google_maps_geocoder.rb]: instead of repeating this 5 times I added some "magic" with active_support inflections help

File: lib/google_maps_geocoder/google_maps_geocoder_error_handler.rb (hunk 5 above)
Comment (kiote) [inline lib/google_maps_geocoder/google_maps_geocoder_error_handler.rb]: yeah, I renamed this and got rid of the module as well, please see second commit

**Comment by ivanoblomov** [comment]
This looks great, @kiote! Just a couple more things though, please. ;)

1. Fix the last [Rubocop warnings](https://travis-ci.org/ivanoblomov/google_maps_geocoder/builds/124783700). It's easy with the `-a` switch.

2. Could you add test coverage for the other error classes you made to `google_maps_geocoder_spec.rb`? Nothing too crazy, just replicate the existing examples.

File: spec/lib/google_maps_geocoder_spec.rb [hunk 10]
```diff
@@ -83,23 +83,16 @@
...
 
   context 'with google returns empty results' do
-    let(:empty_result) { { 'results' => [], 'status' => 'ZERO_RESULTS' } }
-    let(:limit_result) { { 'results' => [], 'status' => 'OVER_QUERY_LIMIT' } }
+    let(:results_hash) { { 'results' => [] } }
 
-    it 'raises an error if we have empty results' do
-      allow_any_instance_of(GoogleMapsGeocoder).to receive(:json_from_url)\
-        .and_return empty_result
+    GoogleMapsGeocoder::ERROR_STATUSES.each do |key, value|
```
Comment (kiote) [inline spec/lib/google_maps_geocoder_spec.rb]: not really sure is it worth or not to make this in a tests, maybe I just need to stay with 5-times repeating code, since it might be easy to read, so if you think it's too much, I can fix it

File: lib/google_maps_geocoder/google_maps_geocoder.rb [hunk 11]
```diff
@@ -27,7 +27,7 @@ def message
 
   class ZeroResultsError < GeocodingError; end
   class QueryLimitError < GeocodingError; end
-  class ReqestDeniedError < GeocodingError; end
+  class RequestDeniedError < GeocodingError; end
```
Comment (kiote) [inline lib/google_maps_geocoder/google_maps_geocoder.rb]: found a typo after testing!

File: google_maps_geocoder.gemspec [hunk 12]
```diff
@@ -21,6 +21,6 @@ Gem::Specification.new do |s|
   s.files       = `git ls-files`.split "\n"
   s.test_files  = `git ls-files -- {test,spec,features}/*`.split "\n"
   s.executables = `git ls-files -- bin/*`.split("\n")
-                  .map { |f| File.basename f }
+                                         .map { |f| File.basename f }
```
Comment (kiote) [inline google_maps_geocoder.gemspec]: this is the result of `rubocop -a` as well

File: spec/lib/google_maps_geocoder_spec.rb (hunk 10 above)
Comment (ivanoblomov) [inline spec/lib/google_maps_geocoder_spec.rb]: This way works great!

**Comment by ivanoblomov** [comment]
Surprised the `-a` didn't fix that last error, but once you do that, I think we're good to merge.

**Comment by kiote** [comment]
Cool! Can you please bump the gem version as well?

**Comment by ivanoblomov** [comment]
Done, let me know if there are any issues and thank for the contribution!

**Comment by kiote** [comment]
Everything looks awesome! 🎆 thanks
